File: 680.valid-palindrome-ii.py
#
# @lc app=leetcode id=680 lang=python3
#
# [680] Valid Palindrome II
#

# @lc code=start
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def is_palindrome(s: str) -> bool:
            return s == s[::-1]
        
        left, right = 0, len(s) - 1
        while left < right:
            if s[left] != s[right]:
                return is_palindrome(s[left:right]) or is_palindrome(s[left+1:right+1])
            left += 1
            right -= 1
        return True
        
# @lc code=end

File: 680.valid-palindrome-ii.kt
/*
 * @lc app=leetcode id=680 lang=kotlin
 *
 * [680] Valid Palindrome II
 */

// @lc code=start
class Solution {
    private fun isPalindrome(s: String, i: Int, j: Int): Boolean {
    var start = i
    var end = j
    while (start < end) {
        if (s[start] != s[end]) {
            return false
        }
        start++
        end--
    }
    return true
    }
    
    fun validPalindrome(s: String): Boolean {
        var i = 0
        var j = s.length - 1
        while (i < j) {
            if (s[i] != s[j]) {
                return isPalindrome(s, i + 1, j) || isPalindrome(s, i, j - 1)

            }
            i++
            j--
        }
        return true
        
    }
}
// @lc code=end

